RunLoop lacks a create operation

kkinnunen-apple opened the pull request:
#### 2b6f2e9cb51ed9c4a01ce76efdab7e1dff3c637b
<pre>
RunLoop lacks a create operation
<a href="https://bugs.webkit.org/show_bug.cgi?id=245673">https://bugs.webkit.org/show_bug.cgi?id=245673</a>
rdar://problem/100411103

Reviewed by Antti Koivisto and Žan Doberšek.

Move existing logic to RunLoop::create(). This makes future new invocations
less error prone.

* Source/WTF/wtf/RunLoop.cpp:
(WTF::RunLoop::create):
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueueBase::platformInitialize):
* Source/WebCore/page/scrolling/ScrollingThread.cpp:
(WebCore::ScrollingThread::ScrollingThread):
(WebCore::ScrollingThread::isCurrentThread):
* Source/WebCore/page/scrolling/ScrollingThread.h:
(WebCore::ScrollingThread::runLoop):
(): Deleted.
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/CompositingRunLoop.cpp:
(WebKit::CompositingRunLoop::CompositingRunLoop):
(WebKit::createRunLoop): Deleted.
* Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/254945@main">https://commits.webkit.org/254945@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dca24bf2dd00932de83a4ef27c724f92a780557c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35260 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99995 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158317 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33758 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28893 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83038 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96394 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96335 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26866 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77503 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26710 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81417 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69738 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34850 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15454 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32662 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16434 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3450 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36428 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39372 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35523 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
```diff
@@ -259,4 +260,34 @@ TEST(WTF_RunLoop, MAYBE_CapabilityIsCurrentNegativeDeathTest)
+        RunLoop::current().stop();
+    });
+    runLoop = nullptr;
+    Util::runFor(.4_s);
```
Maybe it doesn't need to run this long?

dpogue commented on Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
```diff
@@ -41,16 +41,14 @@ WorkQueueBase::WorkQueueBase(RunLoop& runLoop)
         m_threadID = Thread::current().uid();
-#endif
-        m_runLoop = &RunLoop::current();
-        semaphore.signal();
```
Is it intentional that signaling the semaphore is removed, while keeping the wait?

kkinnunen-apple replied:
Thanks!
(Appears that runloop tests aren't run WorkQueueGeneric platforms?)

zdobersek commented on Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
```diff
@@ -41,16 +41,14 @@ WorkQueueBase::WorkQueueBase(RunLoop& runLoop)
-        semaphore.signal();
-        m_runLoop->run();
-    }, ThreadType::Unknown, qos)->detach();
+    });
```
Semaphore's not signaled.

kkinnunen-apple replied:
Thanks!
(Appears that work queue tests aren't run WorkQueueGeneric platforms?)

zdobersek replied:
Only layout tests are run on the GTK tester EWS, for some reason.

kkinnunen-apple committed:
RunLoop lacks a create operation
https://bugs.webkit.org/show_bug.cgi?id=245673
rdar://problem/100411103

Reviewed by Antti Koivisto and Žan Doberšek.

Move existing logic to RunLoop::create(). This makes future new invocations
less error prone.

* Source/WTF/wtf/RunLoop.cpp:
(WTF::RunLoop::create):
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueueBase::platformInitialize):
* Source/WebCore/page/scrolling/ScrollingThread.cpp:
(WebCore::ScrollingThread::ScrollingThread):
(WebCore::ScrollingThread::isCurrentThread):
* Source/WebCore/page/scrolling/ScrollingThread.h:
(WebCore::ScrollingThread::runLoop):
(): Deleted.
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/CompositingRunLoop.cpp:
(WebKit::CompositingRunLoop::CompositingRunLoop):
(WebKit::createRunLoop): Deleted.
* Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/254945@main
